docs/hooks.md

```diff
@@ -0,0 +1,108 @@
+# Hooks
+
+Cucumber provides a number of hooks which allow us to run blocks at various points in the Cucumber test cycle. You can define them anywhere in your support or step definition layers, using the methods `Before` and `After`. Hooks are global by default, meaning they run for every scenario in your
+features. If you want them to run for just certain scenarios, you need to tag
+those scenarios and then use a [tagged hook](hooks.md#tagged-hooks).
+
+## Scenario hooks
+
+`Before` hooks will be run before the first step of each scenario. They will run in the same order of which they are registered.
+
+{% codetabs name="Ruby", type="rb" -%}
+Before do
+  # Do something before each scenario.
+end
+{%- language name="Java", type="java" -%}
+  {% raw %}
+  @Before
+  public void beforeCallingScenario() {
+    System.out.println("*********** About to start the scenario.");
+  }
+  {% endraw %}
+{%- endcodetabs %}
+
+or like this:
+
+{% codetabs name="Ruby", type="rb" -%}
+Before do |scenario|
+  # The +scenario+ argument is optional, but if you use it, you can get the title,
+  # description, or name (title + description) of the scenario that is about to be
+  # executed.
+  Rails.logger.debug "Starting scenario: #{scenario.title}"
+end
+{%- endcodetabs %}
+
+`After` hooks will be run after the last step of each scenario, even when there are failing, undefined, pending or skipped steps. They will run in the opposite order of which they are registered.
+{% codetabs name="Ruby", type="rb" -%}
+After do |scenario|
+  # Do something after each scenario.
+  # The +scenario+ argument is optional, but
+  # if you use it, you can inspect status with
+  # the #failed?, #passed? and #exception methods.
+
+  if scenario.failed?
+    subject = "[Project X] #{scenario.exception.message}"
+    send_failure_email(subject)
+  end
+end
+{%- language name="Java", type="java" -%}
+  {% raw %}
+  @After
+  public void afterRunningScenario(Scenario scenario) {
+    System.out.println("*********** Just finished running scenario: "
+      + scenario.getStatus());
+  }
+  {% endraw %}
+{%- endcodetabs %}
+
+`Around` hooks give you more power than Before and After hooks, allowing you to
```

Comment:
Around hooks have caused a lot of pain, and it's a feature we're likely to deprecate and eventually remove. It only exists for ruby. I suggest not documenting it (and perpetuating its use).